File: examples/ios/objc/GroupedTableView/TableViewController.m
////////////////////////////////////////////////////////////////////////////
//
// Copyright 2014 Realm Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////

#import "TableViewController.h"
#import <Realm/Realm.h>

// Realm model object
@interface DemoObject : RLMObject
@property NSString *phoneNumber;
@property NSDate   *date;
@property NSString *contactName;
@end

@implementation DemoObject
// None needed
@end

static NSString * const kCellID    = @"cell";
static NSString * const kTableName = @"table";

@interface TableViewController ()

@property (nonatomic, strong) RLMSectionedResults<NSString *, DemoObject *> *sectionedResults;
@property (nonatomic, strong) RLMNotificationToken *notification;

@end

@implementation TableViewController

#pragma mark - View Lifecycle

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Section Titles
    [self setupUI];

    self.sectionedResults = [[DemoObject allObjects] sectionedResultsSortedUsingKeyPath:@"contactName"
                                                                              ascending:YES
                                                                               keyBlock:^(DemoObject *object) {
        return [object.contactName substringToIndex:1];
    }];

    // Set realm notification block
    __weak typeof(self) weakSelf = self;
    self.notification = [self.sectionedResults addNotificationBlock:^(RLMSectionedResults<NSString *, DemoObject *> *col,
                                                                      RLMSectionedResultsChange *changes) {
        if (changes) {
            [weakSelf.tableView performBatchUpdates:^{
                [weakSelf.tableView deleteRowsAtIndexPaths:changes.deletions withRowAnimation:UITableViewRowAnimationAutomatic];
                [weakSelf.tableView insertRowsAtIndexPaths:changes.insertions withRowAnimation:UITableViewRowAnimationAutomatic];
                [weakSelf.tableView reloadRowsAtIndexPaths:changes.modifications withRowAnimation:UITableViewRowAnimationAutomatic];
                [weakSelf.tableView insertSections:changes.sectionsToInsert withRowAnimation:UITableViewRowAnimationAutomatic];
                [weakSelf.tableView deleteSections:changes.sectionsToRemove withRowAnimation:UITableViewRowAnimationAutomatic];
            } completion:^(BOOL finished) {
                // Noop
            }];
        }
    }];

    [self.tableView reloadData];
}

#pragma mark - UI

- (void)setupUI
{
    self.title = @"GroupedTableView";
    self.navigationItem.leftBarButtonItem =
    [[UIBarButtonItem alloc] initWithTitle:@"BG Add"
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(backgroundAdd)];
    self.navigationItem.rightBarButtonItem =
    [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
                                                  target:self
                                                  action:@selector(add)];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.sectionedResults.count;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return self.sectionedResults[section].key;
}

- (NSArray<NSString *> *)sectionIndexTitlesForTableView:(UITableView *)tableView
{
    return self.sectionedResults.allKeys;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.sectionedResults[section].count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kCellID];

    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle
                                      reuseIdentifier:kCellID];
    }

    DemoObject *object = self.sectionedResults[indexPath.section][indexPath.row];
    cell.textLabel.text = [NSString stringWithFormat:@"%@: %@", object.contactName, object.phoneNumber];
    cell.detailTextLabel.text = object.date.description;

    return cell;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        RLMRealm *realm = RLMRealm.defaultRealm;
        [realm beginWriteTransaction];
        DemoObject *object = self.sectionedResults[indexPath.section][indexPath.row];
        [realm deleteObject:object];
        [realm commitWriteTransaction];
    }
}

#pragma mark - Actions

- (void)backgroundAdd
{
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    // Import many items in a background thread
    dispatch_async(queue, ^{
        // Get new realm and table since we are in a new thread
        @autoreleasepool {
            RLMRealm *realm = [RLMRealm defaultRealm];
            [realm beginWriteTransaction];
            for (NSInteger index = 0; index < 5; index++) {
                // Add row via dictionary. Order is ignored.
                [DemoObject createInRealm:realm withValue:@{@"phoneNumber": [self randomContactInfo],
                                                             @"date": [NSDate date],
                                                             @"contactName": [self randomContactName]}];
            }
            [realm commitWriteTransaction];
        }
    });
}

- (void)add
{
    [[RLMRealm defaultRealm] transactionWithBlock:^{
        [DemoObject createInDefaultRealmWithValue:@[[self randomContactInfo], [NSDate date], [self randomContactName]]];
    }];
}

#pragma - Helpers

- (NSInteger)randomNumberBetween:(NSInteger)min maxNumber:(NSInteger)max
{
    return min + arc4random_uniform((uint32_t)(max - min + 1));
}

- (NSString *)randomContactInfo
{
    NSInteger rand1 = [self randomNumberBetween:0 maxNumber:9];
    NSInteger rand2 = [self randomNumberBetween:0 maxNumber:9];
    NSInteger rand3 = [self randomNumberBetween:0 maxNumber:9];
    return [NSString stringWithFormat:@"555-55%ld-%ld%ld55", (long)rand1, rand2, rand3];
}

- (NSString *)randomContactName
{
    NSArray *names = @[@"John", @"Mary", @"Fred", @"Sarah", @"Sally", @"James"];
    return [names objectAtIndex:arc4random()%[names count]];
}

@end

File: examples/ios/swift/GroupedTableView/TableViewController.swift
////////////////////////////////////////////////////////////////////////////
//
// Copyright 2014 Realm Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////

import UIKit
import RealmSwift

class DemoObject: Object {
    @Persisted var phoneNumber: String
    @Persisted var date: Date
    @Persisted var contactName: String
    var firstLetter: String {
        guard let char = contactName.first else {
            return ""
        }
        return String(char)
    }
}

class Cell: UITableViewCell {
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String!) {
        super.init(style: .subtitle, reuseIdentifier: reuseIdentifier)
    }

    required init(coder: NSCoder) {
        fatalError("NSCoding not supported")
    }
}

class TableViewController: UITableViewController {
    var notificationToken: NotificationToken?
    var realm: Realm!
    var sectionedResults: SectionedResults<String, DemoObject>!

    override func viewDidLoad() {
        super.viewDidLoad()

        setupUI()
        realm = try! Realm()
        sectionedResults = realm.objects(DemoObject.self)
            .sectioned(by: \.firstLetter, ascending: true)

        // Set realm notification block
        notificationToken = sectionedResults.observe { change in
            switch change {
            case .initial:
                break
            case let .update(_,
                             deletions: deletions,
                             insertions: insertions,
                             modifications: modifications,
                             sectionsToInsert: sectionsToInsert,
                             sectionsToDelete: sectionsToDelete):
                self.tableView.performBatchUpdates {
                    self.tableView.deleteRows(at: deletions, with: .automatic)
                    self.tableView.insertRows(at: insertions, with: .automatic)
                    self.tableView.reloadRows(at: modifications, with: .automatic)
                    self.tableView.insertSections(sectionsToInsert, with: .automatic)
                    self.tableView.deleteSections(sectionsToDelete, with: .automatic)
                }
            }
        }
        tableView.reloadData()
    }

    // UI

    func setupUI() {
        tableView.register(Cell.self, forCellReuseIdentifier: "cell")

        self.title = "GroupedTableView"
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "BG Add", style: .plain, target: self, action: #selector(TableViewController.backgroundAdd))
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(TableViewController.add))
    }

    // Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return sectionedResults.count
    }

    override func sectionIndexTitles(for tableView: UITableView) -> [String]? {
        return sectionedResults.allKeys
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return sectionedResults[section].key
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return sectionedResults[section].count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! Cell

        let object = sectionedResults[indexPath]
        cell.textLabel?.text = "\(object.contactName): \(object.phoneNumber)"
        cell.detailTextLabel?.text = object.date.description

        return cell
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == .delete {
            try! realm.write {
                realm.delete(sectionedResults[indexPath])
            }
        }
    }

    // MARK: Actions

    @objc func backgroundAdd() {
        // Import many items in a background thread
        DispatchQueue.global().async {
            // Get new realm and table since we are in a new thread
            autoreleasepool {
                let realm = try! Realm()
                realm.beginWrite()
                for _ in 0..<5 {
                    // Add row via dictionary. Order is ignored.
                    realm.create(DemoObject.self, value: ["contactName": randomName(), "date": NSDate(), "phoneNumber": randomPhoneNumber()])
                }
                try! realm.commitWrite()
            }
        }
    }

    @objc func add() {
        try! realm.write {
            realm.create(DemoObject.self, value: ["contactName": randomName(), "date": NSDate(), "phoneNumber": randomPhoneNumber()])
        }
    }
}

// MARK: Helpers

func randomPhoneNumber() -> String {
    return "555-55\(Int.random(in: 0...9))5-55\(Int.random(in: 0...9))"
}

func randomName() -> String {
    return ["John", "Jane", "Mary", "Eric", "Sarah", "Sally"].randomElement()!
}
